### ISSUE TITLE
Headers not removed on error pages

### ISSUE THREAD

**Issue by HazCod** [mholt/caddy#1183]
### 1. What version of Caddy are you running (`caddy -version`)?

Caddy 0.9.3
### 2. What are you trying to do?

Remove the Server header at all times.
### 3. What is your entire Caddyfile?

``` text
hazcod.com {

    log stdout
    errors stderr

#    log /tmp/caddy.log {
#        rotate {
#            size 10
#            keep 5
#        }
#    }

#    errors {
#        log /tmp/caddy.log {
#            size 10
#            keep 5
#        }
#    }

    header / {
        "Strict-Transport-Security" "max-age=63072000; includeSubdomains; preload"
        "X-Frame-Options" "DENY"
        "X-Content-Type-Options" "nosniff"
        "X-XSS-Protection" "1; mode=block"
        "Access-Control-Allow-Origin" "{host}"
        "Access-Control-Allow-Methods" "GET, POST, OPTIONS"
        -Server
    }

    gzip
    minify

    fastcgi / php-fpm:8080 php {
        except /css/,/js/,/img/,/fonts/
    }

    expires {
        match .js$ 1w
        match .css$ 1w
        match .png$ 1w
    }
    root /assets/
}

```
### 4. How did you run Caddy (give the full command and describe the execution environment)?

sudo ./caddy -conf Caddyfile -email hazcod@xxx.xx -agree
### 5. What did you expect to see?

No `Server: Caddy` header when receiving a 404
### 6. What did you see instead (give full error messages and/or log)?

`Server: Caddy` on a 404
### 7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?

`header -Server`, query non-existant file and check response headers.

Edit: This is also the case when querying existing directories it seems.

**Comment by tw4452852** [comment]
This problem is due to the header middleware only influences the following chain, while the `Server` header is set at the start of chain. There is quick fix for this case: 

```
diff --git a/caddyhttp/header/header.go b/caddyhttp/header/header.go
index dff925d..18deaa2 100644
--- a/caddyhttp/header/header.go
+++ b/caddyhttp/header/header.go
@@ -104,6 +104,9 @@ func (rww *responseWriterWrapper) addHeader(key, value string) {
 
 // delHeader registers a http.Header.Del operation
 func (rww *responseWriterWrapper) delHeader(key string) {
+       // remove the existing one if any
+       rww.w.Header().Del(key)
+
        rww.ops = append(rww.ops, func(h http.Header) {
                h.Del(key)
        })
```

But it can't cover the all cases: for example, we can't delete [this header](https://github.com/mholt/caddy/blob/master/caddyhttp/httpserver/server.go#L405) which is set from [here](https://github.com/mholt/caddy/blob/master/caddyhttp/httpserver/server.go#L219).
If we want:

```
header / -X-Content-Type-Options
```

@mholt Do you have any good ideas?

**Comment by mholt** [comment]
Hmm, tricky. I don't at this point, but I'll think on it when I have a chance.

**Comment by mholt** [comment]
What if we move the `header` middleware to go before `errors`? Headers doesn't produce errors so we should be good, right?

**Comment by tw4452852** [comment]
No. Because `DefaultErrorFunc` may be called [when all the middlewares return](https://github.com/mholt/caddy/blob/master/caddyhttp/httpserver/server.go#L219) which its `http.ResponseWriter` is the original one not [our wrapper](https://github.com/mholt/caddy/blob/master/caddyhttp/header/header.go#L65). Although we could make a wrapper at [here](https://github.com/mholt/caddy/blob/master/caddyhttp/httpserver/server.go#L210).

**Comment by mholt** [comment]
That's true -- it feels hacky to put the wrapper out there though. Hmm...

**Comment by seglberg** [comment]
Just wanted to point out that this issue also occurs when another middleware short-circuits the chain and instantly returns a response, as in the case with the cors plugin with an OPTIONS preflight. In this case, the Server header is not removed, even with `-Server`, likely due to the same reasons stated above.

**Comment by tw4452852** [comment]
Hi @seglberg , you could work around this by this:

```
- return 200, nil
+ w.WriteHeader(200)
+ return 0, nil                                         
```

**Comment by tw4452852** [comment]
I have just provided a fix. It will handle most cases correctly. But not for the headers added when `header.ServeHTTP` returns.

**Comment by seglberg** [comment]
Thanks for the quick responses. @tw4452852 your suggested work around did in-fact work!

**Comment by mholt** [comment]
Should be fixed in #1212.

**Comment by mholt** [comment]
Whoops, actually, no it's not. But I'll move `header` directive up in the list right now, and push that and that should fix it.
